sky/packages/sky/lib/base/gesture_manager.dart

```diff
@@ -0,0 +1,165 @@
+  }
+
+  void _attemptToChooseRecognizer(sky.Event event) {
+    assert(_choosenRecognizer == null);
+    _forEachRecognizer((GestureRecognizer recognizer, GestureState state) {
+      if (_choosenRecognizer != null)
+        return;
+      GestureDisposition disposition = recognizer.observeEvent(state, event);
+      if (disposition == GestureDisposition.accepted) {
+        _choosenRecognizer = recognizer;
+      } else if (disposition == GestureDisposition.rejected) {
+        recognizer.notifyListeners(state, GestureDisposition.rejected);
+        _recognizers.remove(recognizer);
+      }
+    });
+    if (_choosenRecognizer == null && _recognizers.length == 1)
+      _choosenRecognizer = _recognizers.keys.first;
+    if (_choosenRecognizer == null)
+      return;
+    _choosenGestureState = _recognizers[_choosenRecognizer];
+    _forEachRecognizer((GestureRecognizer recognizer, GestureState state) {
+      if (recognizer == _choosenRecognizer)
+        return;
+      recognizer.notifyListeners(state, GestureDisposition.rejected);
+    });
+    _recognizers.clear();
+  }
+
+  void _notifyRecognizers(sky.Event event) {
+    if (_choosenRecognizer != null) {
+      assert(_recognizers.isEmpty);
+      assert(_choosenGestureState != null);
+      _choosenRecognizer.notifyListeners(_choosenGestureState, GestureDisposition.accepted);
+    } else {
+      _forEachRecognizer((GestureRecognizer recognizer, GestureState state) {
+        recognizer.notifyListeners(state, GestureDisposition.possible);
+      });
+    }
+  }
+
+  void handleEvent(sky.Event event) {
+    if (_choosenRecognizer == null) {
+      _attemptToChooseRecognizer(event);
+    } else {
+      _choosenRecognizer.observeEvent(_choosenGestureState, event);
+    }
+    _notifyRecognizers(event);
+  }
+}
+
+class GestureManager extends HitTestTarget {
+  final Map<int, PointerState> _pointers = new Map<int, PointerState>();
+
+  GestureSubscription detectGesture(int pointer, GestureRecognizer recognizer, GestureListener listener) {
+    assert(pointer != null);
+    assert(recognizer != null);
+    assert(listener != null);
+    PointerState pointerState = _pointers.putIfAbsent(pointer, () => new PointerState());
+    return pointerState.detectGesture(recognizer, listener);
+  }
+
+  EventDisposition handleEvent(sky.Event event, HitTestEntry entry) {
+    int pointer = _getPointer(event);
+    if (pointer == null)
+      return EventDisposition.ignored;
+    PointerState pointerState = _pointers[pointer];
+    if (pointerState == null)
+      return EventDisposition.ignored;
+    pointerState.handleEvent(event);
+    // TODO(abarth): When should we remove pointerState? It's tempting to remove
+    // it when we see the pointerup, but we still might get gesture events from
+    // the platform recognizer that we need to process. We probably need some
+    // event that marks the end of the gesture packet.
+    return EventDisposition.processed;
```

Comment:
You can't remove it on pointerUp, because otherwise gestures like double-tap couldn't work on devices where a single pointer can go down twice (e.g. some Samsung devices with hover touch). You have to remove it when all the gestures that were registered for the pointer have said they are finished and ready for a new gesture.